mapit_labour/management/commands/mapit_labour_import_addressbase_core.py

```diff
@@ -44,40 +44,30 @@ def handle(self, *args, **kwargs):
         super().handle(*args, **kwargs)
 
     def handle_rows(self, csv: DictReader):
-        for row in csv:
-            self.handle_row(row)
+        if self.purge:
+            UPRN.objects.all().delete()
 
-            if self.count["total"] % self.often == 0:
+            for _, uprns in groupby(
+                (self.create_uprn(row) for row in csv),
+                lambda _: self.count["total"] // self.batch_size,
+            ):
+                with transaction.atomic():
+                    UPRN.objects.bulk_create(uprns)
                 self.print_stats()
         self.print_stats()
 
-    def handle_row(self, row: Dict[str, str]):
+    def create_uprn(self, row: Dict[str, str]):
         row = {k.lower(): v for k, v in row.items()}
-        e = float(row["easting"])
-        n = float(row["northing"])
-        location = Point(e, n, srid=27700)
-        postcode = row["postcode"].replace(" ", "")
-        uprn = row["uprn"]
 
-        if self.purge:
-            UPRN.objects.create(
-                uprn=uprn,
-                postcode=postcode,
-                location=location,
-                addressbase=row,
-            )
-            self.count["created"] += 1
-        else:
-            _, created = UPRN.objects.update_or_create(
-                uprn=uprn,
-                defaults=dict(
-                    postcode=postcode,
-                    location=location,
-                    addressbase=row,
-                ),
-            )
-            self.count["created" if created else "updated"] += 1
         self.count["total"] += 1
+        self.count["created"] += 1
```

Comment:
Does it need to have both total and created? The stats output doesn't seem accurate now.

Reply:
Now we can update existing records this is more useful. Total is number of rows processed from the CSV, and is a sum of created and updated.